lightning/src/chain/channelmonitor.rs

```diff
@@ -1915,35 +1767,34 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
+		let revk_outp = RevokedOutput::build(per_commitment_point, self.counterparty_tx_cache.counterparty_delayed_payment_base_key, self.counterparty_tx_cache.counterparty_htlc_base_key, per_commitment_key, tx.output[0].value, self.counterparty_tx_cache.on_counterparty_tx_csv);
+		let justice_package = PackageTemplate::build_package(htlc_txid, 0, PackageSolvingData::RevokedOutput(revk_outp), height + self.counterparty_tx_cache.on_counterparty_tx_csv as u32, true, 0, None, height);
+		let claimable_outpoints = vec!(justice_package);
 		let outputs = vec![(0, tx.output[0].clone())];
 		(claimable_outpoints, Some((htlc_txid, outputs)))
 	}
 
-	// Returns (1) `ClaimRequest`s that can be given to the OnChainTxHandler, so that the handler can
+	// Returns (1) `PackageTemplate`s that can be given to the OnChainTxHandler, so that the handler can
 	// broadcast transactions claiming holder HTLC commitment outputs and (2) a holder revokable
 	// script so we can detect whether a holder transaction has been seen on-chain.
-	fn get_broadcasted_holder_claims(&self, holder_tx: &HolderSignedTx) -> (Vec<ClaimRequest>, Option<(Script, PublicKey, PublicKey)>) {
+	fn get_broadcasted_holder_claims(&self, holder_tx: &HolderSignedTx, height: u32) -> (Vec<PackageTemplate>, Option<(Script, PublicKey, PublicKey)>) {
 		let mut claim_requests = Vec::with_capacity(holder_tx.htlc_outputs.len());
 
 		let redeemscript = chan_utils::get_revokeable_redeemscript(&holder_tx.revocation_key, self.on_holder_tx_csv, &holder_tx.delayed_payment_key);
 		let broadcasted_holder_revokable_script = Some((redeemscript.to_v0_p2wsh(), holder_tx.per_commitment_point.clone(), holder_tx.revocation_key.clone()));
 
 		for &(ref htlc, _, _) in holder_tx.htlc_outputs.iter() {
 			if let Some(transaction_output_index) = htlc.transaction_output_index {
-				claim_requests.push(ClaimRequest { absolute_timelock: ::std::u32::MAX, aggregable: false, outpoint: BitcoinOutPoint { txid: holder_tx.txid, vout: transaction_output_index as u32 },
-					witness_data: InputMaterial::HolderHTLC {
-						preimage: if !htlc.offered {
-								if let Some(preimage) = self.payment_preimages.get(&htlc.payment_hash) {
-									Some(preimage.clone())
-								} else {
-									// We can't build an HTLC-Success transaction without the preimage
-									continue;
-								}
-							} else { None },
-						amount: htlc.amount_msat,
-				}});
+				let htlc_output = HolderHTLCOutput::build(if !htlc.offered {
+					if let Some(preimage) = self.payment_preimages.get(&htlc.payment_hash) {
+						Some(preimage.clone())
+					} else {
+						// We can't build an HTLC-Success transaction without the preimage
+						continue;
+					}
+				} else { None }, htlc.amount_msat);
+				let htlc_package = PackageTemplate::build_package(holder_tx.txid, transaction_output_index, PackageSolvingData::HolderHTLCOutput(htlc_output), height, false, 0, None, height);
```

Comment:
I'm a bit confused by the timelock semantics. Previously we were always setting `absolute_timelock` to none here, but now we set it to the current height (though it should be current height + 1 if we're trying to do the "always timelock transactions" thing). Should we be moving towards `absolute_timelock` really meaning "the real timelock when this transaction can hit the chain"? I do think that would make it much more useful.

In that case, does the absolute_timelock here not need to be HTLC timelock + height (for HTLC-Timeout transactions)? May be worth adding some debug assertions that the transactions generated have a timelock that matches the absolute_timelock in the PackageTemplate.

Reply:
I think I would agree on defining `absolute_timelock` as "the real hit when counterparty transaction can hit the chain" ?

Right, timelock should not be height but in fact `htlc.cltv_expiry`. Though current value of `height` is just a tweak to avoid generating non-bumped HTLC-transactions for nothing, which would break some tests. Note, it should be changed in the rest of the patchset iirc.

Reply:
Right, I think the other few calls already use that definition, just not this one, though it may also make sense to always do max(HTLC timeout, current height + 1)?

Reply:
IIRC, if HTLC timeout is already behind the chain tip, we're going to bump it at every block according to `get_block_height_timer`, so doing max(HTLC timeout, current height + 1) isn't going to change anything.